src/Bicep.Core/TypeSystem/DeployTimeConstantVisitor.cs

```diff
@@ -0,0 +1,195 @@
+                            }
+                        }
+                        break;
+                }
+            }
+        }
+
+        public override void VisitObjectSyntax(ObjectSyntax syntax)
+        {
+            if (this.bodyObj == null)
+            {
+                return;
+            }
+
+            foreach (var deployTimeIdentifier in ObjectSyntaxExtensions.ToNamedPropertyDictionary(syntax))
+            {
+                if (this.bodyObj.Properties.TryGetValue(deployTimeIdentifier.Key, out var propertyType) && propertyType.Flags.HasFlag(TypePropertyFlags.DeployTimeConstant))
+                {
+                    this.currentProperty = deployTimeIdentifier.Key;
+                    this.VisitObjectPropertySyntax(deployTimeIdentifier.Value);
+                }
+            }
+        }
+
+        public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
+        {
+            base.VisitObjectPropertySyntax(syntax);
+            // if an error is found at the end we emit it and move on to the next key of this object declaration.
+            if (this.errorSyntax != null)
+            {
+                this.AppendError(this.errorSyntax);
+                this.errorSyntax = null;
+            }
+        }
+
+        public override void VisitPropertyAccessSyntax(PropertyAccessSyntax syntax)
+        {
+            base.VisitPropertyAccessSyntax(syntax);
+            if (this.errorSyntax != null)
+            {
+                // Due to the nature of visitPropertyAccessSyntax, we have to visit every level of this
+                // nested errorSyntax recursively. The last one will be shown to the user.
+                this.errorSyntax = syntax;
+            }
+            else if (syntax.BaseExpression is VariableAccessSyntax variableAccessSyntax)
+            {
+                // validate only on resource and module symbols
+                var baseSymbol = model.GetSymbolInfo(variableAccessSyntax);
+                switch (baseSymbol)
+                {
+                    case ResourceSymbol:
+                    case ModuleSymbol:
+                        if (TypeAssignmentVisitor.UnwrapType(((DeclaredSymbol)baseSymbol).Type) is ObjectType referencedBodyObj)
+                        {
+                            var property = syntax.PropertyName.IdentifierName;
+                            if (referencedBodyObj.Properties.TryGetValue(property, out var propertyType) &&
+                            !propertyType.Flags.HasFlag(TypePropertyFlags.DeployTimeConstant))
+                            {
+                                this.errorSyntax = syntax;
+                                this.referencedBodyObj = referencedBodyObj;
+                            }
+                        }
+                        break;
+                }
+            }
+        }
+        private void AppendError(SyntaxBase syntax)
+        {
+            if (this.currentProperty == null)
```

Comment:
you can use `is null` instead `== null`

Reply:
I will confer with the team and possibly explore the idea of making this a project wide change in a separate PR (most of bicep tends to use the operators instead of the keywords). Thanks for flagging this!